### PR TITLE
Prevent macOS fullscreen crash

### PR DESCRIPTION
**Pull request by FreezyLemon**
See [this comment](https://github.com/ppy/osu-framework/pull/1588#issuecomment-394136207).

This needs to be tested as I don't have a Mac, but it _should_ fix the crash.

### CODE REVIEW

**Comment by FreezyLemon** [comment]
From discord, for reference:
This seems to work, but another crash occurs [one line further down](https://github.com/ppy/opentk/blob/netstandard/src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs#L1020).
```
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'nextEventMatchingMask should only be called from the Main Thread!'
```

Still investigating that one, but looking at stackoverflow questions about this it seems to be pretty self-explanatory.

File: src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs [hunk 1]
```diff
@@ -1376,18 +1376,25 @@ private void SetCursorGrab(bool shouldGrab)
             }
         }
 
+        private NSApplicationPresentationOptions lastAutoHideOptions;
+
         private void SetMenuVisible(bool visible)
         {
             var options = (NSApplicationPresentationOptions)Cocoa.SendInt(NSApplication.Handle, selPresentationOptions);
             var changedOptions = NSApplicationPresentationOptions.HideMenuBar | NSApplicationPresentationOptions.HideDock;
```
Comment (VPeruS) [inline src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs]: make `static const HideMenuAndDock` in class

File: src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs [hunk 2]
```diff
@@ -1376,18 +1376,25 @@ private void SetCursorGrab(bool shouldGrab)
...
 
             if (!visible)
             {
+                lastAutoHideOptions = options & (NSApplicationPresentationOptions.AutoHideMenuBar | NSApplicationPresentationOptions.AutoHideDock);
```
Comment (VPeruS) [inline src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs]: for this two AutoHideMenuAndDock

File: src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs [hunk 3]
```diff
@@ -1376,18 +1376,25 @@ private void SetCursorGrab(bool shouldGrab)
...
+
                 options |= changedOptions;
```
Comment (VPeruS) [inline src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs]: this is noop

**Comment by VPeruS** [comment]
why you didn't make options &= ~(NSApplicationPresentationOptions.AutoHideMenuBar | NSApplicationPresentationOptions.AutoHideDock) at start of function? And just left everyting like it was before?

**Comment by FreezyLemon** [comment]
@VPeruS 
Let's say that a window has the `AutoHideDock` flag set.
When we switch to fullscreen, we have to disable this so we can "manually" activate `HideDock`.
But when switching back out of fullscreen, we want to re-apply the same flags that we had when switching _into_ fullscreen.

**Comment by peppy** [comment]
Still crashes here. When entering or exiting full screen (also when exiting borderless).

**Comment by FreezyLemon** [comment]
@peppy got a stack trace? I'll try to to fix the remaining stuff (see: second post) today or at least over the next few days...

Or did the crash mentioned in OP happen again?

As I said, I can't test this myself, so I really need all the information I can get..

**Comment by VPeruS** [comment]
https://github.com/opentk/opentk/blob/9bd713fb084fad3a4fc8137bae2eaa36b06e70e7/src/OpenTK/DisplayDevice.cs#L194 looks like this *must* have return

**Comment by FreezyLemon** [comment]
I think I'll give up on the threading issue for now. I don't think I have the necessary expertise, time or equipment (still can't test it myself) to fix this, really. This would be one thing to do on Windows, but on Mac? I'll have to pass.

The remaining change in this PR is still valid, though.

**Comment by swoolcock** [comment]
I've had a bit of a look at this, and I managed to get it to "kind of" work by deferring the entire `WindowState` setter to an `Action` queue that gets run in `ProcessEvents`.

That being said, Apple recommend NOT changing the display resolution when going fullscreen.  I think it would be better to just set a logical resolution and get OpenGL to handle it automatically.

https://developer.apple.com/library/archive/documentation/GraphicsImaging/Conceptual/OpenGL-MacProgGuide/EnablingOpenGLforHighResolution/EnablingOpenGLforHighResolution.html#//apple_ref/doc/uid/TP40001987-CH1001-SW12

> You should avoid changing the display mode of the system.

**Comment by peppy** [comment]
A good start would be to make this work without changing the screen resolution. It doesn't even work in that scenario currently.

**Comment by swoolcock** [comment]
In reference to https://github.com/ppy/osu-framework/pull/1691

I'm going to do this the "right way" by fixing it in OpenTK instead.  There's a bit of a problem though in that while the "force the resolution" method for `WindowState.Fullscreen` works (mostly), that's not really what we should be doing in macOS.

I will make a separate pull request when it's done, and we can discuss from there.

**Comment by peppy** [comment]
Should I not be reviewing ppy/osu-framework#1691 in its current state, then?

**Comment by swoolcock** [comment]
Correct, leave it for now.

**Comment by peppy** [comment]
Is this one no longer relevant?

**Comment by FreezyLemon** [comment]
Not really, unless I missed anything.

The initial issue was that switching to fullscreen (and borderless too, IIRC) would crash the game on macOS. @swoolcock seems to have fixed that in https://github.com/ppy/osu-framework/pull/1691, so this isn't really needed anymore.

**Pull request closed without merging**